internal/querynode/scheduler_policy.go

```diff
@@ -2,23 +2,305 @@ package querynode
 
 import (
 	"container/list"
+	"container/ring"
+	"time"
+
+	"github.com/milvus-io/milvus/internal/util/contextutil"
+)
+
+var (
+	_ scheduleReadPolicy = &fifoScheduleReadPolicy{}
+	_ scheduleReadPolicy = &userTaskPollingScheduleReadPolicy{}
 )
 
-type scheduleReadTaskPolicy func(sqTasks *list.List, targetUsage int32, maxNum int32) ([]readTask, int32)
+type scheduleReadPolicy interface {
+	// Read task count inside.
+	len() int
+
+	// Add a new task into ready task.
+	addTask(rt readTask)
+
+	// Merge a new task into exist ready task.
+	// Return true if merge success.
+	mergeTask(rt readTask) bool
+
+	// Schedule new task to run.
+	schedule(targetCPUUsage int32, maxNum int32) ([]readTask, int32)
+}
+
+// Create a new schedule policy.
+func newReadScheduleTaskPolicy(policyName string) scheduleReadPolicy {
+	switch policyName {
+	case "":
+		fallthrough
+	case "fifo":
```

Comment:
It is recommended to define "fifo" and "user-task-polling" as Const variables under the querynode package.